### PR TITLE
Alerting: Add support for UTF-8 characters in notification policies and silences

### LINKED ISSUE #81308: Use double quotes when creating notification policies or silences containing reserved characters
When creating notification policies or silences that contain one of the reserved characters:

```
{}!=~,\"'`
```

then the text should be double quoted before being sent to the API.

For example, the following screenshot does not contain reserved characters:

image

However, this one uses all reserved characters:

image

It cannot be sent to the Alertmanager as:

```
foo={}!=~,\"'`
```

but must instead be sent as:

```
foo="{}!=~,\\\"'`"
```

Here the right hand side has been double quoted, and both the `\` and the double quote `"` have been escaped as `\\` and `\"`.

#### More examples

`foo=bar` can be kept as `foo=bar`, however `foo=bar!` should be sent as `foo="bar!"`.

It is also OK to double quote regardless of whether it contains reserved characters or not. For example, `foo=bar` is identical to `foo="bar"`.

This is needed for the rollout of UTF-8 in Alertmanager in Grafana Managed Alerts, Grafana Cloud Alertmanager and Prometheus Alertmanager.

### PR DESCRIPTION
**Pull request by konrad147**
**What is this feature?**
This PR adds UTF-8 characters support to the Notification Policies and Silences UI

**Why do we need this feature?**
We need this feature to be compatible with the recent changes in Alertmanager API

**Who is this feature for?**
All users

Fixes #81308 

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/alerting/unified/utils/matchers.ts [hunk 1]
```diff
@@ -108,4 +108,31 @@ export const normalizeMatchers = (route: Route): ObjectMatcher[] => {
...
+    return input;
+  }
+
+  return input
+    .replace(/^"(.*)"$/, '$1')
+    .replace(/\\"/g, '"')
+    .replace(/\\\\/g, '\\');
+}
+
+export const matcherFormatter = {
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/utils/matchers.ts:127]: 🙌 love this approach

File: public/app/features/alerting/unified/components/notification-policies/EditNotificationPolicyForm.tsx [hunk 2]
```diff
@@ -134,7 +135,9 @@ export const AmRoutesExpandedForm = ({
                               error={errors.object_matchers?.[index]?.value?.message}
                             >
                               <Input
-                                {...register(`object_matchers.${index}.value`, { required: 'Field is required' })}
+                                {...register(`object_matchers.${index}.value`, {
+                                  required: 'Field is required',
+                                })}
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/components/notification-policies/EditNotificationPolicyForm.tsx:140]: only prettier changes, right?

**Review by gillesdemey — APPROVED** [review]
LGTM just a few nits

File: public/app/features/alerting/unified/utils/matchers.ts [hunk 3]
```diff
@@ -108,4 +108,31 @@ export const normalizeMatchers = (route: Route): ObjectMatcher[] => {
...
+  default: ([name, operator, value]: ObjectMatcher): string => {
+    return `${name} ${operator} ${value}`;
+  },
+  cloud: ([name, operator, value]: ObjectMatcher): string => {
```
[the remark this answers is not in the document]
Comment (konrad147) [reply public/app/features/alerting/unified/utils/matchers.ts:131]: Good point!

**Comment by grobinson-grafana** [comment]
I think we need to make the Value field optional, otherwise it's not possible to write `foo=""`. If we put `""` into the Value field, it becomes `foo="\"\""`. This is correct, we just need to make the field non-required to allow no value.

**Comment by gillesdemey** [comment]
Good catch! An empty value should be allowed for [negative matchers](https://github.com/grafana/grafana/pull/77292)

**Review by gillesdemey — APPROVED** [review]

**Comment by grobinson-grafana** [comment]
Looks great to me! Well done 🥳 

Screenshot 2024-02-09 at 2 18 25 PM
Screenshot 2024-02-09 at 2 18 34 PM

**Pull request merged**